File: src/lib/autopush.ts
// ScanWise auto-push loop — runs inside the Next.js server process.
// Calls runAutoPushCycle() every 30s. Logs to scripts/autopush.log.
// Every git op is wrapped in try/catch so a single failure never kills the loop.

import { execFileSync } from "node:child_process";
import { appendFileSync, mkdirSync } from "node:fs";
import { dirname, resolve } from "node:path";

const REPO = resolve(process.cwd());
const LOG_PATH = resolve(REPO, "scripts/autopush.log");
const POLL_INTERVAL_MS = 30_000;
const BRANCH = "main";

function log(message: string): void {
  const ts = new Date().toISOString();
  const line = `${ts} ${message}\n`;
  try {
    mkdirSync(dirname(LOG_PATH), { recursive: true });
    appendFileSync(LOG_PATH, line, "utf8");
  } catch {
    // ignore — never crash the server on log failure
  }
  // also echo to server stdout for visibility
  process.stdout.write(`[autopush] ${line}`);
}

function runGit(args: string[]): { ok: boolean; out: string } {
  try {
    const out = execFileSync("git", args, {
      cwd: REPO,
      encoding: "utf8",
      timeout: 120_000,
      stdio: ["ignore", "pipe", "pipe"],
    });
    return { ok: true, out };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return { ok: false, out: msg };
  }
}

function runAutoPushCycle(): void {
  // 1. Fetch remote (best-effort)
  runGit(["fetch", "origin", BRANCH]);

  // 2. Rebase any remote changes
  runGit(["pull", "--rebase", "--autostash", "origin", BRANCH]);

  // 3. Check for pending changes
  const status = runGit(["status", "--porcelain"]);
  if (!status.ok) {
    log(`ERROR: git status failed: ${status.out}`);
    return;
  }
  const pending = status.out.trim();
  if (!pending) return; // nothing to do

  // 4. Stage everything not gitignored
  runGit(["add", "-A"]);

  // 5. Build summary
  const diffStat = runGit(["diff", "--cached", "--stat"]);
  const summary =
    diffStat.out.trim().split("\n").pop() || "(no diff)";

  // 6. Commit
  const ts = new Date().toISOString();
  const commitMsg = `auto: ${ts} — ${summary}`;
  const commit = runGit(["commit", "-m", commitMsg, "--quiet"]);
  if (!commit.ok) {
    log(`INFO: commit failed (likely nothing to commit): ${commit.out}`);
    return;
  }

  // 7. Push
  const push = runGit(["push", "origin", BRANCH]);
  if (push.ok) {
    const sha = runGit(["rev-parse", "--short", "HEAD"]).out.trim();
    log(`pushed @ ${sha}: ${summary}`);
  } else {
    log(`ERROR: push failed: ${push.out}`);
  }
}

let started = false;

export function startAutoPush(): void {
  if (started) return; // guard against double-registration in dev HMR
  started = true;

  log(`=== ScanWise autopush started inside Next.js server (PID ${process.pid}) ===`);
  log(`poll interval: ${POLL_INTERVAL_MS}ms | branch: ${BRANCH}`);

  // Run one cycle immediately so we don't wait 30s for the first push.
  try {
    runAutoPushCycle();
  } catch (err) {
    log(`initial cycle error: ${err}`);
  }

  // Then poll forever.
  setInterval(() => {
    try {
      runAutoPushCycle();
    } catch (err) {
      log(`cycle error (continuing): ${err}`);
    }
  }, POLL_INTERVAL_MS);
}

File: scripts/autopush.py
#!/usr/bin/env python3
"""ScanWise auto-push daemon.

Polls the working tree every 30s. If anything changed, commits with a
timestamped message and pushes to origin/main.

Robustness:
  - Single-instance via fcntl.flock on this very file.
  - Never dies on git failures (every git op is wrapped in try/except).
  - Logs every cycle to scripts/autopush.log.
  - Handles SIGTERM / SIGHUP cleanly.
"""

from __future__ import annotations

import fcntl
import os
import subprocess
import sys
import time
from datetime import datetime, timezone
from pathlib import Path

REPO = Path("/home/z/my-project")
LOG_PATH = REPO / "scripts" / "autopush.log"
POLL_INTERVAL = 30  # seconds
BRANCH = "main"


def log(msg: str) -> None:
    ts = datetime.now(timezone.utc).isoformat()
    line = f"{ts} {msg}"
    try:
        with LOG_PATH.open("a", encoding="utf-8") as fh:
            fh.write(line + "\n")
    except OSError:
        pass
    print(line, flush=True)


def run_git(args: list[str]) -> tuple[int, str, str]:
    """Run a git command, return (returncode, stdout, stderr)."""
    try:
        proc = subprocess.run(
            ["git", *args],
            cwd=REPO,
            capture_output=True,
            text=True,
            timeout=120,
        )
        return proc.returncode, proc.stdout, proc.stderr
    except Exception as exc:  # noqa: BLE001
        return -1, "", str(exc)


def main() -> int:
    # Single-instance guard: flock on this script file
    try:
        lock_fh = open(__file__, "r")
        try:
            fcntl.flock(lock_fh, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BlockingIOError:
            log("another autopush is already running; exiting")
            return 0
    except OSError as exc:
        log(f"could not open self for flock: {exc}")
        return 1

    log(f"=== ScanWise autopush daemon started (PID {os.getpid()}) ===")
    log(f"poll interval: {POLL_INTERVAL}s | branch: {BRANCH}")

    while True:
        try:
            # 1. Fetch remote refs (best-effort)
            run_git(["fetch", "origin", BRANCH])

            # 2. Rebase any remote changes (with autostash)
            run_git(["pull", "--rebase", "--autostash", "origin", BRANCH])

            # 3. Check for pending changes
            rc, porcelain, _ = run_git(["status", "--porcelain"])
            pending = porcelain.strip()

            if pending:
                # Stage everything not gitignored
                run_git(["add", "-A"])

                # Build a short summary
                _, summary_raw, _ = run_git(["diff", "--cached", "--stat"])
                summary = summary_raw.strip().splitlines()[-1] if summary_raw.strip() else "(no diff)"
                timestamp = datetime.now(timezone.utc).isoformat()

                # Commit
                run_git(["commit", "-m", f"auto: {timestamp} — {summary}", "--quiet"])

                # Push
                rc, out, err = run_git(["push", "origin", BRANCH])
                if rc == 0:
                    short_sha = run_git(["rev-parse", "--short", "HEAD"])[1].strip()
                    log(f"pushed @ {short_sha}: {summary}")
                else:
                    err_short = (err or out or "").strip().splitlines()[-1] if (err or out) else ""
                    log(f"ERROR: push failed (rc={rc}): {err_short}")
            else:
                # Nothing to do this cycle — don't log to avoid noise
                pass

        except Exception as exc:  # noqa: BLE001
            log(f"cycle error (continuing): {exc}")

        time.sleep(POLL_INTERVAL)


if __name__ == "__main__":
    sys.exit(main())
